File: tests/python/test_download_sources.py
"""
Tests for v2.2.0 occurrence download scripts and predictor download scripts.
No live API calls — all tests use mock data and pure logic.
Run: pytest tests/python/test_download_sources.py -v
"""
import sys
import importlib.util
import types
from pathlib import Path
import tempfile
from unittest.mock import MagicMock

import pandas as pd
import pytest

SCRIPTS = Path(__file__).parents[2] / "skills" / "ecological-data-foundation" / "scripts"
GEO_SCRIPTS = Path(__file__).parents[2] / "skills" / "geoprocessing-for-ecology" / "scripts"

# ── Standard schema columns ───────────────────────────────────────────────────
STANDARD_SCHEMA = [
    "species", "decimalLatitude", "decimalLongitude", "eventDate",
    "countryCode", "basisOfRecord", "coordinateUncertaintyInMeters",
    "datasetName", "occurrenceID", "source", "download_doi",
]


def make_stub(name: str):
    mod = types.ModuleType(name)
    mod.__spec__ = MagicMock()
    mod.get_observations = MagicMock(return_value={"results": [], "total_results": 0})
    return mod


def inject_stubs(*names: str) -> None:
    for name in names:
        if name not in sys.modules:
            sys.modules[name] = make_stub(name)


def load_module(script_path: Path, mod_name: str):
    inject_stubs(
        "pyinaturalist", "robis", "rredlist", "requests",
        "cdsapi", "pystac_client", "stackstac", "planetary_computer",
        "rasterio", "rasterio.mask", "shapely", "shapely.wkt",
    )
    spec = importlib.util.spec_from_file_location(mod_name, script_path)
    mod  = importlib.util.module_from_spec(spec)
    original_exit = sys.exit
    sys.exit = lambda code=0: None  # prevent sys.exit(1) from killing test collection
    try:
        spec.loader.exec_module(mod)
    finally:
        sys.exit = original_exit
    return mod


# ── Load modules once ─────────────────────────────────────────────────────────
inat_mod  = load_module(SCRIPTS / "download_from_inat.py",  "download_inat")
obis_mod  = load_module(SCRIPTS / "download_from_obis.py",  "download_obis")
iucn_mod  = load_module(SCRIPTS / "download_from_iucn.py",  "download_iucn")
ebird_mod = load_module(SCRIPTS / "download_from_ebird.py", "download_ebird")
pred_mod  = load_module(GEO_SCRIPTS / "download_predictors.py", "download_pred")


# ─────────────────────────────────────────────────────────────────────────────
# 1. Schema validation helpers
# ─────────────────────────────────────────────────────────────────────────────

class TestStandardSchema:
    """Verify that standardise_records functions produce the required column set."""

    def test_inat_standardise_produces_schema(self):
        obs = [{
            "location": "-5.0,-55.0",
            "observed_on": "2022-03-15",
            "place_guess": "Brazil",
            "id": 12345678,
            "positional_accuracy": 10,
        }]
        df = inat_mod.standardise_records(obs, "Panthera onca")
        missing = [c for c in STANDARD_SCHEMA if c not in df.columns]
        assert missing == [], f"Missing schema columns: {missing}"

    def test_inat_standardise_drops_missing_coords(self):
        obs = [
            {"location": "invalid", "observed_on": "2022-01-01", "place_guess": "", "id": 1, "positional_accuracy": None},
            {"location": "-5.0,-55.0", "observed_on": "2022-01-01", "place_guess": "BR", "id": 2, "positional_accuracy": 10},
        ]
        df = inat_mod.standardise_records(obs, "Panthera onca")
        assert len(df) == 1
        assert df.iloc[0]["occurrenceID"] == "2"

    def test_inat_source_column_is_inat(self):
        obs = [{"location": "-5.0,-55.0", "observed_on": "2022-01-01",
                "place_guess": "BR", "id": 99, "positional_accuracy": 5}]
        df = inat_mod.standardise_records(obs, "Myrmecophaga tridactyla")
        assert df.iloc[0]["source"] == "iNaturalist"

    def test_obis_standardise_produces_schema(self):
        records = [{
            "decimalLatitude": -5.0,
            "decimalLongitude": -55.0,
            "eventDate": "2020-06-01",
            "countryCode": "BR",
            "basisOfRecord": "HUMAN_OBSERVATION",
            "coordinateUncertaintyInMeters": 100,
            "datasetName": "OBIS Dataset",
            "occurrenceID": "urn:obis:1234",
            "depth": 0.0,
        }]
        df = obis_mod.standardise_records(records, "Chelonia mydas")
        missing = [c for c in STANDARD_SCHEMA if c not in df.columns]
        assert missing == [], f"Missing schema columns: {missing}"

    def test_obis_extra_columns_present(self):
        records = [{"decimalLatitude": -5.0, "decimalLongitude": -55.0,
                    "depth": 10.0, "occurrenceID": "x1"}]
        df = obis_mod.standardise_records(records, "Chelonia mydas")
        assert "depth" in df.columns
        assert "marine" in df.columns

    def test_obis_source_column_is_obis(self):
        records = [{"decimalLatitude": -5.0, "decimalLongitude": -55.0}]
        df = obis_mod.standardise_records(records, "Chelonia mydas")
        assert df.iloc[0]["source"] == "OBIS"

    def test_iucn_standardise_produces_schema_no_countries(self):
        assessment = {
            "result": [{
                "taxonid": 15951,
                "category": "NT",
                "criteria": "A2cd",
                "population_trend": "Decreasing",
                "assessment_date": "2018-01-01",
            }]
        }
        df = iucn_mod.standardise_records(assessment, [], "Panthera onca")
        missing = [c for c in STANDARD_SCHEMA if c not in df.columns]
        assert missing == [], f"Missing schema columns: {missing}"

    def test_iucn_standardise_with_countries(self):
        assessment = {
            "result": [{"taxonid": 15951, "category": "NT", "criteria": "",
                        "population_trend": "Decreasing", "assessment_date": "2018-01-01"}]
        }
        countries = [{"code": "BR"}, {"code": "CO"}, {"code": "PE"}]
        df = iucn_mod.standardise_records(assessment, countries, "Panthera onca")
        assert len(df) == 3
        assert set(df["countryCode"]) == {"BR", "CO", "PE"}

    def test_iucn_extra_columns_present(self):
        assessment = {"result": [{"taxonid": 1, "category": "LC", "criteria": "",
                                   "population_trend": "Stable", "assessment_date": "2020-01-01"}]}
        df = iucn_mod.standardise_records(assessment, [], "Thalurania furcata")
        assert "rl_category" in df.columns
        assert "assessment_year" in df.columns


# ─────────────────────────────────────────────────────────────────────────────
# 2. OBIS quality flag filtering
# ─────────────────────────────────────────────────────────────────────────────

class TestOBISQualityFlags:

    def test_no_coord_flag_removed(self):
        records = [
            {"decimalLatitude": -5.0, "decimalLongitude": -55.0, "flags": "NO_COORD"},
            {"decimalLatitude": -5.0, "decimalLongitude": -55.0, "flags": ""},
        ]
        filtered = obis_mod.apply_quality_flags(records)
        assert len(filtered) == 1

    def test_zero_coord_flag_removed(self):
        records = [{"decimalLatitude": 0.0, "decimalLongitude": 0.0, "flags": "ZERO_COORD"}]
        filtered = obis_mod.apply_quality_flags(records)
        assert len(filtered) == 0

    def test_on_land_flag_removed(self):
        records = [{"decimalLatitude": -5.0, "decimalLongitude": -55.0, "flags": "ON_LAND"}]
        filtered = obis_mod.apply_quality_flags(records)
        assert len(filtered) == 0

    def test_clean_records_pass_through(self):
        records = [
            {"decimalLatitude": -5.0, "decimalLongitude": -55.0, "flags": ""},
            {"decimalLatitude": -6.0, "decimalLongitude": -56.0, "flags": None},
        ]
        filtered = obis_mod.apply_quality_flags(records)
        assert len(filtered) == 2


# ─────────────────────────────────────────────────────────────────────────────
# 3. eBird EBD parsing helpers
# ─────────────────────────────────────────────────────────────────────────────

class TestEBirdParsing:

    def _make_ebd_df(self, species="Jabiru mycteria", protocol="Stationary",
                      approved="1", year=2021, lat=-5.0, lon=-55.0):
        return pd.DataFrame({
            "SCIENTIFIC NAME":           [species],
            "LATITUDE":                  [lat],
            "LONGITUDE":                 [lon],
            "OBSERVATION DATE":          [f"{year}-06-15"],
            "COUNTRY CODE":              ["BR"],
            "SAMPLING EVENT IDENTIFIER": ["S12345678"],
            "PROTOCOL TYPE":             [protocol],
            "EFFORT DISTANCE KM":        [1.5],
            "DURATION MINUTES":          [60],
            "OBSERVER ID":               ["obsr123"],
            "APPROVED":                  [approved],
            "REVIEWED":                  ["0"],
        })

    def test_standardise_ebd_schema(self):
        df = self._make_ebd_df()
        std = ebird_mod.standardise_ebd(df, "Jabiru mycteria")
        missing = [c for c in STANDARD_SCHEMA if c not in std.columns]
        assert missing == [], f"Missing: {missing}"

    def test_standardise_ebd_extra_columns(self):
        df = self._make_ebd_df()
        std = ebird_mod.standardise_ebd(df, "Jabiru mycteria")
        assert "effort_distance_km" in std.columns
        assert "duration_minutes"   in std.columns
        assert "observer_id"        in std.columns

    def test_standardise_ebd_source_is_ebird(self):
        df = self._make_ebd_df()
        std = ebird_mod.standardise_ebd(df, "Jabiru mycteria")
        assert std.iloc[0]["source"] == "eBird"

    def test_standardise_ebd_wrong_species_empty(self):
        df = self._make_ebd_df(species="Jabiru mycteria")
        std = ebird_mod.standardise_ebd(df, "Ara ararauna")
        assert len(std) == 0


# ─────────────────────────────────────────────────────────────────────────────
# 4. Metadata save functions (no network; writes to temp dir)
# ─────────────────────────────────────────────────────────────────────────────

class TestMetadataSave:

    def test_inat_metadata_written(self, tmp_path):
        inat_mod.save_metadata(tmp_path, "Panthera onca", 42, 2010, 2023, "research")
        meta_file = tmp_path / "download_metadata_iNat_Panthera_onca.txt"
        assert meta_file.exists()
        content = meta_file.read_text(encoding="utf-8")
        assert "iNaturalist" in content
        assert "42" in content

    def test_obis_metadata_written(self, tmp_path):
        obis_mod.save_metadata(tmp_path, "Chelonia mydas", 15, 1990, 2022, None)
        meta_file = tmp_path / "download_metadata_OBIS_Chelonia_mydas.txt"
        assert meta_file.exists()
        content = meta_file.read_text(encoding="utf-8")
        assert "OBIS" in content
        assert "CC0" in content

    def test_iucn_metadata_written(self, tmp_path):
        iucn_mod.save_metadata(tmp_path, "Panthera onca", 15951, "NT", "2018")
        meta_file = tmp_path / "download_metadata_IUCN_Panthera_onca.txt"
        assert meta_file.exists()
        content = meta_file.read_text(encoding="utf-8")
        assert "IUCN" in content
        assert "15951" in content


# ─────────────────────────────────────────────────────────────────────────────
# 5. Predictor download helpers (no network)
# ─────────────────────────────────────────────────────────────────────────────

class TestPredictorHelpers:

    def test_sha256_nonexistent_returns_na(self, tmp_path):
        result = pred_mod.sha256_file(tmp_path / "nonexistent.tif")
        assert result == "N/A"

    def test_sha256_real_file(self, tmp_path):
        f = tmp_path / "test.bin"
        f.write_bytes(b"hello world")
        result = pred_mod.sha256_file(f)
        assert len(result) == 64  # SHA256 hex digest length

    def test_save_metadata_empty_rows(self, tmp_path):
        pred_mod.save_metadata(tmp_path, [])
        # Should not create file (warning only)
        assert not (tmp_path / "predictor_metadata.csv").exists()

    def test_save_metadata_writes_csv(self, tmp_path):
        rows = [{"source": "WorldClim_v2.1", "variable": "BIO1",
                 "resolution": "2.5m", "file": "/tmp/bio1.tif",
                 "citation": "Fick & Hijmans (2017)",
                 "licence": "CC BY 4.0", "download_date": "2026-03-05"}]
        pred_mod.save_metadata(tmp_path, rows)
        meta_path = tmp_path / "predictor_metadata.csv"
        assert meta_path.exists()
        df = pd.read_csv(meta_path)
        assert len(df) == 1
        assert df.iloc[0]["source"] == "WorldClim_v2.1"

    def test_valid_resolutions_dict_contains_key(self):
        assert 2.5 in pred_mod.VALID_RESOLUTIONS
        assert 10  in pred_mod.VALID_RESOLUTIONS

    def test_chelsa_base_url_format(self):
        # Ensure the base URL is correct (no live request)
        assert "chelsa" in pred_mod.CHELSA_BASE.lower()
        assert pred_mod.CHELSA_BASE.startswith("https://")

    def test_worldclim_base_url_format(self):
        assert "worldclim" in pred_mod.WORLDCLIM_BASE.lower()
        assert pred_mod.WORLDCLIM_BASE.startswith("https://")

File: tests/r/test-download-sources.R
# Tests for v2.2.0 occurrence download scripts (schema + argument validation).
# No live API calls — all tests use mock data and pure logic.
# Run: Rscript tests/r/test-download-sources.R

library(testthat)
library(dplyr)
library(readr)

# ── Standard schema columns ───────────────────────────────────────────────────
STANDARD_SCHEMA <- c(
  "species", "decimalLatitude", "decimalLongitude", "eventDate",
  "countryCode", "basisOfRecord", "coordinateUncertaintyInMeters",
  "datasetName", "occurrenceID", "source", "download_doi"
)

# ── Helper: build a mock occurrence data.frame matching the standard schema ───
make_mock_occ <- function(n = 5, source = "GBIF") {
  data.frame(
    species                        = rep("Panthera onca", n),
    decimalLatitude                = runif(n, -10, 0),
    decimalLongitude               = runif(n, -70, -50),
    eventDate                      = paste0("2020-0", seq_len(n), "-01"),
    countryCode                    = "BR",
    basisOfRecord                  = "HUMAN_OBSERVATION",
    coordinateUncertaintyInMeters  = sample(100:5000, n, replace = TRUE),
    datasetName                    = paste0(source, " Dataset"),
    occurrenceID                   = paste0(source, ":", seq_len(n)),
    source                         = source,
    download_doi                   = NA_character_,
    stringsAsFactors               = FALSE
  )
}

# ─────────────────────────────────────────────────────────────────────────────
# 1. Standard schema column presence
# ─────────────────────────────────────────────────────────────────────────────

test_that("mock_occ has all standard schema columns", {
  df <- make_mock_occ()
  missing <- setdiff(STANDARD_SCHEMA, names(df))
  expect_equal(missing, character(0), info = paste("Missing:", paste(missing, collapse = ", ")))
})

test_that("iNaturalist schema includes all standard columns", {
  # Simulate output from download_from_inat.R standardisation block
  df <- data.frame(
    species                        = "Myrmecophaga tridactyla",
    decimalLatitude                = -5.0,
    decimalLongitude               = -55.0,
    eventDate                      = "2021-08-10",
    countryCode                    = "Brazil",
    basisOfRecord                  = "HUMAN_OBSERVATION",
    coordinateUncertaintyInMeters  = 10,
    datasetName                    = "iNaturalist",
    occurrenceID                   = "12345678",
    source                         = "iNaturalist",
    download_doi                   = NA_character_,
    stringsAsFactors               = FALSE
  )
  missing <- setdiff(STANDARD_SCHEMA, names(df))
  expect_equal(missing, character(0))
})

test_that("OBIS schema includes all standard columns plus depth and marine", {
  df <- data.frame(
    species                        = "Chelonia mydas",
    decimalLatitude                = -5.0,
    decimalLongitude               = -35.0,
    eventDate                      = "2019-03-01",
    countryCode                    = "BR",
    basisOfRecord                  = "HUMAN_OBSERVATION",
    coordinateUncertaintyInMeters  = NA_real_,
    datasetName                    = "OBIS Dataset",
    occurrenceID                   = "urn:obis:123",
    source                         = "OBIS",
    download_doi                   = NA_character_,
    depth                          = 0.0,
    marine                         = TRUE,
    stringsAsFactors               = FALSE
  )
  missing <- setdiff(STANDARD_SCHEMA, names(df))
  expect_equal(missing, character(0))
  expect_true("depth"  %in% names(df))
  expect_true("marine" %in% names(df))
})

test_that("IUCN schema includes standard columns plus rl_category", {
  df <- data.frame(
    species                        = "Panthera onca",
    decimalLatitude                = NA_real_,
    decimalLongitude               = NA_real_,
    eventDate                      = NA_character_,
    countryCode                    = "BR",
    basisOfRecord                  = "LITERATURE",
    coordinateUncertaintyInMeters  = NA_real_,
    datasetName                    = "IUCN Red List",
    occurrenceID                   = "IUCN:15951:BR",
    source                         = "IUCN",
    download_doi                   = NA_character_,
    rl_category                    = "NT",
    rl_criteria                    = "A2cd",
    population_trend               = "Decreasing",
    assessment_year                = 2018L,
    stringsAsFactors               = FALSE
  )
  missing <- setdiff(STANDARD_SCHEMA, names(df))
  expect_equal(missing, character(0))
  expect_true("rl_category"     %in% names(df))
  expect_true("assessment_year" %in% names(df))
})

test_that("eBird schema includes standard columns plus effort columns", {
  df <- data.frame(
    species                        = "Jabiru mycteria",
    decimalLatitude                = -10.0,
    decimalLongitude               = -55.0,
    eventDate                      = "2021-09-01",
    countryCode                    = "BR",
    basisOfRecord                  = "HUMAN_OBSERVATION",
    coordinateUncertaintyInMeters  = NA_real_,
    datasetName                    = "eBird Basic Dataset",
    occurrenceID                   = "S98765432",
    source                         = "eBird",
    download_doi                   = NA_character_,
    effort_distance_km             = 1.5,
    duration_minutes               = 60,
    observer_id                    = "obsr123",
    stringsAsFactors               = FALSE
  )
  missing <- setdiff(STANDARD_SCHEMA, names(df))
  expect_equal(missing, character(0))
  expect_true("effort_distance_km" %in% names(df))
  expect_true("duration_minutes"   %in% names(df))
  expect_true("observer_id"        %in% names(df))
})

# ─────────────────────────────────────────────────────────────────────────────
# 2. Coordinate validity checks
# ─────────────────────────────────────────────────────────────────────────────

test_that("records with missing lat are removed in schema filtering", {
  df <- make_mock_occ(5)
  df$decimalLatitude[2] <- NA
  df_clean <- df[!is.na(df$decimalLatitude) & !is.na(df$decimalLongitude), ]
  expect_equal(nrow(df_clean), 4)
})

test_that("records with missing lon are removed", {
  df <- make_mock_occ(3)
  df$decimalLongitude[1] <- NA
  df_clean <- df[!is.na(df$decimalLatitude) & !is.na(df$decimalLongitude), ]
  expect_equal(nrow(df_clean), 2)
})

test_that("valid coordinates are retained", {
  df <- make_mock_occ(10)
  df_clean <- df[!is.na(df$decimalLatitude) & !is.na(df$decimalLongitude), ]
  expect_equal(nrow(df_clean), 10)
})

# ─────────────────────────────────────────────────────────────────────────────
# 3. OBIS quality flag logic (pure R, no API)
# ─────────────────────────────────────────────────────────────────────────────

apply_obis_qflags <- function(df) {
  bad_flags <- c("NO_COORD", "ZERO_COORD", "ON_LAND", "DEPTH_EXCEEDS_BATH")
  if (!"flags" %in% names(df)) return(df)
  keep <- !grepl(paste(bad_flags, collapse = "|"), df$flags, ignore.case = TRUE)
  df[keep, ]
}

test_that("NO_COORD flag removes record", {
  df <- data.frame(flags = c("NO_COORD", ""), stringsAsFactors = FALSE)
  result <- apply_obis_qflags(df)
  expect_equal(nrow(result), 1)
})

test_that("ON_LAND flag removes record", {
  df <- data.frame(flags = c("ON_LAND", "valid"), stringsAsFactors = FALSE)
  result <- apply_obis_qflags(df)
  expect_equal(nrow(result), 1)
})

test_that("clean records pass through quality filter", {
  df <- data.frame(flags = c("", "", ""), stringsAsFactors = FALSE)
  result <- apply_obis_qflags(df)
  expect_equal(nrow(result), 3)
})

test_that("DEPTH_EXCEEDS_BATH flag removes record", {
  df <- data.frame(flags = c("DEPTH_EXCEEDS_BATH"), stringsAsFactors = FALSE)
  result <- apply_obis_qflags(df)
  expect_equal(nrow(result), 0)
})

# ─────────────────────────────────────────────────────────────────────────────
# 4. Source column values
# ─────────────────────────────────────────────────────────────────────────────

test_that("source column for GBIF records is GBIF", {
  df <- make_mock_occ(n = 1, source = "GBIF")
  expect_equal(df$source, "GBIF")
})

test_that("source column for iNaturalist records is iNaturalist", {
  df <- make_mock_occ(n = 1, source = "iNaturalist")
  expect_equal(df$source, "iNaturalist")
})

test_that("source column for OBIS records is OBIS", {
  df <- make_mock_occ(n = 1, source = "OBIS")
  expect_equal(df$source, "OBIS")
})

# ─────────────────────────────────────────────────────────────────────────────
# 5. Script file existence checks
# ─────────────────────────────────────────────────────────────────────────────

BASE_SCRIPTS <- file.path(
  dirname(dirname(dirname(normalizePath(sys.frames()[[1]]$ofile, mustWork = FALSE)))),
  "skills", "ecological-data-foundation", "scripts"
)
GEO_BASE <- file.path(
  dirname(dirname(dirname(normalizePath(sys.frames()[[1]]$ofile, mustWork = FALSE)))),
  "skills", "geoprocessing-for-ecology", "scripts"
)

test_that("download_from_gbif.R exists", {
  expect_true(file.exists(file.path(BASE_SCRIPTS, "download_from_gbif.R")))
})
test_that("download_from_inat.R exists", {
  expect_true(file.exists(file.path(BASE_SCRIPTS, "download_from_inat.R")))
})
test_that("download_from_ebird.R exists", {
  expect_true(file.exists(file.path(BASE_SCRIPTS, "download_from_ebird.R")))
})
test_that("download_from_obis.R exists", {
  expect_true(file.exists(file.path(BASE_SCRIPTS, "download_from_obis.R")))
})
test_that("download_from_iucn.R exists", {
  expect_true(file.exists(file.path(BASE_SCRIPTS, "download_from_iucn.R")))
})
test_that("download_predictors.R exists", {
  expect_true(file.exists(file.path(GEO_BASE, "download_predictors.R")))
})

# ─────────────────────────────────────────────────────────────────────────────
# 6. Script header compliance (Usage: on line 1)
# ─────────────────────────────────────────────────────────────────────────────

check_usage_line1 <- function(script_path) {
  line1 <- readLines(script_path, n = 1, warn = FALSE)
  grepl("^# Usage:", line1)
}

test_that("download_from_inat.R has Usage: on line 1", {
  expect_true(check_usage_line1(file.path(BASE_SCRIPTS, "download_from_inat.R")))
})
test_that("download_from_ebird.R has Usage: on line 1", {
  expect_true(check_usage_line1(file.path(BASE_SCRIPTS, "download_from_ebird.R")))
})
test_that("download_from_obis.R has Usage: on line 1", {
  expect_true(check_usage_line1(file.path(BASE_SCRIPTS, "download_from_obis.R")))
})
test_that("download_from_iucn.R has Usage: on line 1", {
  expect_true(check_usage_line1(file.path(BASE_SCRIPTS, "download_from_iucn.R")))
})
test_that("download_predictors.R has Usage: on line 1", {
  expect_true(check_usage_line1(file.path(GEO_BASE, "download_predictors.R")))
})

# ─────────────────────────────────────────────────────────────────────────────
# 7. suppressPackageStartupMessages present in all new R scripts
# ─────────────────────────────────────────────────────────────────────────────

check_suppress <- function(script_path) {
  lines <- readLines(script_path, warn = FALSE)
  any(grepl("suppressPackageStartupMessages", lines))
}

test_that("download_from_inat.R uses suppressPackageStartupMessages", {
  expect_true(check_suppress(file.path(BASE_SCRIPTS, "download_from_inat.R")))
})
test_that("download_from_ebird.R uses suppressPackageStartupMessages", {
  expect_true(check_suppress(file.path(BASE_SCRIPTS, "download_from_ebird.R")))
})
test_that("download_from_obis.R uses suppressPackageStartupMessages", {
  expect_true(check_suppress(file.path(BASE_SCRIPTS, "download_from_obis.R")))
})
test_that("download_from_iucn.R uses suppressPackageStartupMessages", {
  expect_true(check_suppress(file.path(BASE_SCRIPTS, "download_from_iucn.R")))
})

# ─────────────────────────────────────────────────────────────────────────────
# 8. Inline logger present in all new R scripts
# ─────────────────────────────────────────────────────────────────────────────

check_logger <- function(script_path) {
  lines <- readLines(script_path, warn = FALSE)
  any(grepl("log_step", lines))
}

test_that("download_from_inat.R has log_step", {
  expect_true(check_logger(file.path(BASE_SCRIPTS, "download_from_inat.R")))
})
test_that("download_from_ebird.R has log_step", {
  expect_true(check_logger(file.path(BASE_SCRIPTS, "download_from_ebird.R")))
})
test_that("download_from_obis.R has log_step", {
  expect_true(check_logger(file.path(BASE_SCRIPTS, "download_from_obis.R")))
})
test_that("download_from_iucn.R has log_step", {
  expect_true(check_logger(file.path(BASE_SCRIPTS, "download_from_iucn.R")))
})
test_that("download_predictors.R has log_step", {
  expect_true(check_logger(file.path(GEO_BASE, "download_predictors.R")))
})

# ─────────────────────────────────────────────────────────────────────────────
# 9. Resource files exist
# ─────────────────────────────────────────────────────────────────────────────

BASE_RES <- file.path(
  dirname(dirname(dirname(normalizePath(sys.frames()[[1]]$ofile, mustWork = FALSE)))),
  "skills", "ecological-data-foundation", "resources"
)
GEO_RES <- file.path(
  dirname(dirname(dirname(normalizePath(sys.frames()[[1]]$ofile, mustWork = FALSE)))),
  "skills", "geoprocessing-for-ecology", "resources"
)

test_that("data-citation-guide.md exists", {
  expect_true(file.exists(file.path(BASE_RES, "data-citation-guide.md")))
})
test_that("global-predictor-sources.md exists", {
  expect_true(file.exists(file.path(GEO_RES, "global-predictor-sources.md")))
})

cat("\n✓ All download-sources tests completed.\n")
